[1.21.7][B3D][Breaking] Add GpuDevice configuration, validation, and property reporting

RogueLogix opened the pull request:
- Adds GpuDeviceProperties to allow a backend to report properties as `glGetIneger` may not always be valid to call. Also used by validation layer.
   - Currently only backend info is reported, other info to be added later.
 - Adds GpuDeviceFeatures to allow configuration of features extending what default B3D allows
   - Despite being part of base B3D, `logicOp` is disabled by default due to driver incompatibility issues, with a check for known problematic drivers
 - Adds validation GpuDevice layer in development to catch use of disabled features and usage outside property limits
 
This will allow (some) future additions to B3D to be non-breaking changes as use of non-supported (including previously non-existent) features/enum values is considered invalid behavior that the validation layer should catch in dev.
 
This is a breaking change due to the requirements on mods to enable features they could previously use without enablement (logicOp) and alternate backends must report their properties and fire the configuration event.

RogueLogix committed:
GpuDevice configuration and validation

Technici4n commented on src/client/java/net/neoforged/neoforge/client/ClientHooks.java:
```diff
@@ -1128,4 +1134,12 @@ public static List<PictureInPictureRendererRegistration<?>> gatherPictureInPictu
     }
+
+    public static GpuDevice createGpuDevice(long window, int debugLevel, boolean syncDebug, BiFunction<ResourceLocation, ShaderType, String> defaultShaderSource, boolean enableDebugLabels) {
+        final var glDevice = new NeoGlDevice(window, debugLevel, syncDebug, defaultShaderSource, enableDebugLabels);
```
final var??

RogueLogix replied:
Is this a problem?

Technici4n commented on src/client/java/net/neoforged/neoforge/client/event/ConfigureGpuDeviceEvent.java:
```diff
@@ -0,0 +1,42 @@
+import net.neoforged.neoforge.client.neo3d.GpuDeviceFeatures;
+import net.neoforged.neoforge.client.neo3d.GpuDeviceProperties;
+
+public class ConfigureGpuDeviceEvent extends Event implements IModBusEvent {
```
This needs javadoc.

RogueLogix replied:
resolving in favor of not-outdated one

RogueLogix committed:
mark GpuDeviceFeatures.enableAll and hasAll as @ApiStatus.Internal

RogueLogix committed:
Allow for extensions of ValidationGpuDevice/CommandEncoder/RenderPass

RogueLogix committed:
Validate GpuBuffer and GpuTexture usage bits

RogueLogix committed:
Mark B3D enums as non-exhaustive, and add validation for known values

Technici4n commented on src/main/java/net/neoforged/neoforge/internal/NonExhaustiveEnum.java:
```diff
@@ -0,0 +1,14 @@
+
+@ApiStatus.Internal
+@Target(ElementType.TYPE)
+public @interface NonExhaustiveEnum {}
```
where javadoc?

RogueLogix replied:
added

Technici4n commented on src/client/java/net/neoforged/neoforge/client/loading/NeoForgeLoadingOverlay.java:
```diff
@@ -115,7 +115,11 @@ static class ExternalTexture extends AbstractTexture {
-            this.textureView = RenderSystem.getDevice().createTextureView(texture);
+            var gpuDevice = RenderSystem.getDevice();
+            if (gpuDevice instanceof ValidationGpuDevice) {
+                gpuDevice = ((ValidationGpuDevice) gpuDevice).getRealDevice();
```
Why is this needed? I would expect `createTextureView` to be suitably implemented by `ValidationGpuDevice`.

RogueLogix replied:
The loading overlay has to reach behind the validation device's back to import an external texture directly into the `GlDevice`, and as a result this then needs to do so again here because it doesn't have a `ValidationGpuTexture` instance that the `ValidationGpuDevice` is expecting but rather a `GlGpuTexture` instance.

This will be removed when the loading overlay is updated to allow it to be universal on top of B3D (see my prototype FML `APIWindow`/`APIRenderer`).

Technici4n replied:
Ok fair enough. I would suggest the following though: `if (gpuDevice instanceof ValidationGpuDevice validationDevice)`.

RogueLogix replied:
ahh, yes, will do that

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceProperties.java:
```diff
@@ -0,0 +1,156 @@
+import java.util.Set;
+
+public class GpuDeviceProperties {
+    public final String backendName = defaultVal("Default");
```
What is the point of all of this?

RogueLogix replied:
To report in a uniform way what backend is being used, what API its using, and what version of that API its using without being reliant on building against that backend to do an instanceof check. This is also passed in the configure event which is fired before the `GpuDevice` is available to mods to do a different check against.
`GpuDeviceProperties` itself could be extended to get around that last limitation, but I don't feel that that is a better solution than this.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceFeatures.java:
```diff
@@ -0,0 +1,67 @@
+    @ApiStatus.Internal
+    public boolean hasAll(GpuDeviceFeatures features) {
+        try {
+            for (final var field : fields) {
```
I don't like this usage of reflection.

RogueLogix replied:
It's not my favorite either, but I felt it would be more reasonable than the `if` or `||`/`&&` hell this would otherwise grow to. Currently there is only one field, but it likely wont stay that way for long.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceProperties.java:
```diff
@@ -0,0 +1,156 @@
+     */
+    public final int maximumImageDimensionCube = atLeast(1024);
+
+    public static <T extends GpuDeviceProperties> T create(Class<T> clazz, Map<String, Object> values) {
```
Reflection mess again :/

RogueLogix replied:
Because these are "final" they must be set via reflection, or in the constructor. Because this also needs to be binary compatible as additional fields get added, setting in constructor is not an option.

Technici4n replied:
I think we should go for an interface of getter methods. That will preserve backwards compatibility without weird hacks.

RogueLogix replied:
How would that work for backends extending this class?

XFactHD replied:
I agree with Tech that this should be an interface. Adding new properties would work similarly to adding new feature getters to `GpuDeviceFeatures`, i.e. they get added with a default impl which returns a value that makes the property effectively absent/unsupported (i.e. zero size for a buffer or texture-like thing, etc.).

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/validation/ValidationCommandEncoder.java:
```diff
@@ -0,0 +1,144 @@
+
+    @Override
+    public void clearColorTexture(GpuTexture texture, int clearColor) {
+        assert texture instanceof ValidationGpuTexture;
```
Asserts are typically not enabled, don't use the `assert` keyword.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/event/ConfigureGpuDeviceEvent.java:
```diff
@@ -0,0 +1,42 @@
+        return enabledFeatures.clone();
+    }
+
+    public void requireFeatures(GpuDeviceFeatures requiredFeatures) {
```
I don't think this is a good way to do it. I would much rather have a simple `boolean` getter and setter for each feature. Maybe on an interface rather than the event so we could make it abstract.

RogueLogix replied:
I'm not opposed to a different solution, how exactly would you have your proposed alternative work?

Technici4n replied:
Something like this?
```java
if (event.isLogicOpAvailable()) {
    event.enableLogicOp(); // Throws if not supported
}
```
We could then add extra stuff later with more methods in the event, e.g. for a hypothetical `goodPerformance` feature:
```java
public boolean isGoodPerformanceAvailable() {
    return false;
}
public void enableGoodPerformance() {
    throw new UnsupportedOperationException("Good performance is not available");
}
```
Of course each backend then needs to override these methods in its corresponding event.

We could also go for an intermediate interface such that we can make additions since the last BC window abstract in every new BC window. (Abstract events cannot be listened to, so we can't make the event methods abstract).

RogueLogix replied:
We'd need to go for the intermediate for binary compat. The same 'struct' here is also re-used later to report what is actively enabled, would that be its own thing, or?

Technici4n replied:
No idea, it would be worth experimenting with different approaches. Presumably the interface can be reused for the getters, and the setters can be in the event.

marchermans replied:
Agreed with tech here. Readonly interface, but with setters on the collecting event.

RogueLogix replied:
It is that now (outdated comment)

RogueLogix replied:
@Technici4n is this resolved now? also the related comment for GpuDeviceFeatures and GpuDeviceProperties.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/validation/ValidationGpuDevice.java:
```diff
@@ -0,0 +1,171 @@
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+public class ValidationGpuDevice implements GpuDevice {
```
This whole thing might be better as patches in the actual GL device, no? It would reduce the size of the PR by a lot as I understand it, by making the changes a lot more targeted.

RogueLogix replied:
No, it can't be. This is not targeting validating usage for _just_ the `GlDevice`, but _any_ `GpuDevice` implementation. Having it be its own layer also checks against people reaching behind the interface's back to get to the default `GlDevice`. I also expect it to grow past what would be manageable in patches.

marchermans replied:
Agreed this makes sense to have seperate. The PR is larger, but maintanance wise this should be easier to track as new methods added to GpuDevice by mojang should trigger a compiler error.

Technici4n reviewed (changes requested):
Feels way too complicated to me but I might be missing something.

RogueLogix committed:
Switch GpuDeviceFeatures to an interface

XFactHD commented on src/client/java/net/neoforged/neoforge/client/neo3d/opengl/NeoGlDeviceFeatures.java:
```diff
@@ -0,0 +1,57 @@
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+public class NeoGlDeviceFeatures implements GpuDeviceFeatures {
```
I would prefer if this was split into two separate classes:
- `MutableGpuDeviceFeatures` which is only used for enabled features, is backend-independent and takes another, pre-populated `GpuDeviceFeatures` instance for the available features
- `NeoGlDeviceFeatures` (i.e. a backend-specific impl of `GpuDeviceFeatures`) which collects and holds the available features and is passed to the former

Only the former would then have the setters for enabling features, would be exposed as the mutable type in the configuration event and later as the immutable type on the `GpuDevice` (can still add the locking if you're worried about people downcasting it).

RogueLogix replied:
if it would just be adding a "Mutable" interface even if the impl doesn't change, sure, I don't have a super strong opinion either way.

XFactHD replied:
The specific setup I have in mind would look like this:
```java
interface GpuDeviceFeatures {
    boolean logicOp();
}
class MutableGpuDeviceFeatures implements GpuDeviceFeatures {
    private final GpuDeviceFeatures availableFeatures;
    private boolean logicOp = false;
    public MutableGpuDeviceFeatures(GpuDeviceFeatures availableFeatures) {
        this.availableFeatures = availableFeatures;
    }
    @Override
    public boolean logicOp() {
        return logicOp;
    }
    public void enableLogicOp() {
        if (availableFeatures.logicOp()) throw new UnsupportedOperationException("LogicOp not supported");
        logicOp = true;
    }
}
class GlDeviceFeatures implements GpuDeviceFeatures {
    private final boolean logicOp;
    public GlDeviceFeatures() {
        // Check caps, OS, etc.
        this.logicOp = ...;
    }
    @Override
    public boolean logicOp() {
        return logicOp;
    }
}
```

XFactHD commented on patches/com/mojang/blaze3d/opengl/GlDevice.java.patch:
```diff
@@ -1,9 +1,20 @@
  
+ @OnlyIn(Dist.CLIENT)
+-public class GlDevice implements GpuDevice {
++public abstract class GlDevice implements GpuDevice {
```
I would prefer the `GpuDeviceProperties` and `GpuDeviceFeatures` for the default GL device to be patched into `GlDevice` directly instead of patching it to be abstract and then extending it.
If the `GpuDeviceFeatures` setup is moved to a `ClientHooks` method and the `GpuDeviceProperties` are changed to an interface whose implementation does the property collection, then the patch size would stay reasonable as well.

RogueLogix replied:
the main problem i see with the route of being in patches is not with this specifically but the future, and how much of a PITA patches are to write by comparison (porting maintenance aside).

As a specific example. adding a vertexOffset/firstInstance capable draw command would end up at call to `glDrawElementsInstancedBaseVertexBaseInstanceARB` which is already bad enough, it really doesn't need to be `org.lwjgl.opengl.ARBBaseInstance.glDrawElementsInstancedBaseVertexBaseInstanceARB` because its in a patch. 
Apply to anything and everything else and I'd much prefer to have the neo extensions be implemented in an actual extension class instead. Indirect would be similar, SSBOs could be similar, proper texture view would be unreasonable in patches, compute shaders will be unreasonable in patches.
Doing this could be punted to the future (as this specific case isn't really problematic), but I still think that extending GlDevice/GlCommandEncoder/GlRenderPass still the better route longer term. If extending is done for any of them, they would eventually end up marked as abstract

Technici4n replied:
I would also prefer that the new methods be moved to an extension interface `GpuDeviceExtension`, with implementations patched into `GlDevice` directly.

RogueLogix replied:
moved to GlDevice

RogueLogix committed:
Update Features and Properties to be fetch only interfaces with immutable records.
Feature enablement is now done directly on configuration event.

marchermans reviewed (changes requested):
General consensus here:
The idea is good. The reflection and execution less so. I tend to incline with the ideas of @Technici4n and @XFactHD here. The readonly interface with mutable setters is the better approach.

marchermans commented on src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceProperties.java:
```diff
@@ -5,152 +5,213 @@
-    public final String apiVersionString = defaultVal("3.2");
-    public final int apiVersionMajor = defaultVal(3);
-    public final int apiVersionMinor = defaultVal(2);
+public interface GpuDeviceProperties {
```
Missing Documentation on this type, in this area of the code documentation is extremely important, so it would be nice if all methods would be properly documented as to their contract.

marchermans commented on patches/com/mojang/blaze3d/platform/DepthTestFunction.java.patch:
```diff
@@ -0,0 +1,10 @@
+ 
+ @OnlyIn(Dist.CLIENT)
+ @DontObfuscate
++@net.neoforged.neoforge.internal.NonExhaustiveEnum
```
Add a single line as to why this is non exhaustive.

RogueLogix replied:
Added, to all NonExhaustiveEnum annotations

marchermans commented on patches/com/mojang/blaze3d/platform/DestFactor.java.patch:
```diff
@@ -0,0 +1,10 @@
+ 
+ @OnlyIn(Dist.CLIENT)
+ @DontObfuscate
++@net.neoforged.neoforge.internal.NonExhaustiveEnum
```
Add a single line as to why this is non exhaustive

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceFeatures.java:
```diff
@@ -0,0 +1,32 @@
+     */
+    boolean logicOp();
+
+    record Immutable(boolean logicOp) implements GpuDeviceFeatures {
+        /**
+         * Immutable record should only be constructed from an instance of the interface
+         * <br>
+         * Additional elements may be added without being considered a breaking change
+         *
+         * @see Immutable#Immutable(GpuDeviceFeatures)
+         */
+        @ApiStatus.Internal
+        public Immutable {}
+
+        public Immutable(GpuDeviceFeatures features) {
+            this(features.logicOp());
+        }
+    }
```
Why is this in the API? It's trivial for each backend to replicate.

RogueLogix replied:
currently its trivial, it wont stay that way. the one for properties is already not _completely_ trivial.

~~i haven't done this yet, but the calls in GpuDevice should be changed to return these records (currently returning the interfaces) so that in most cases the vtable lookup can be skipped.~~ i didn't do this because then extension interfaces wont work because records cant be extended, forgot about that.

Technici4n replied:
What you can also do is add a `GpuDeviceFeatures.copyOf()` static method such that the record can be kept package-private and thus internal.

RogueLogix replied:
id actually probably make it non-static, so that implementations can decide where to forward it exactly, but similar end result.

ie: Cinnbar may want to forward this to its own record that also has its own extended features in it, and that ensures all copies (in case mods decode to make their own) will have the full feature set still.

Technici4n replied:
Well, modders are not supposed to be using it directly, and I'd rather not have it as an interface method. It's just meant to be convenience for backends to save a bit of code, so if it starts to make the design any more complex we should just scrap it. And maybe we should, to avoid copying features not supported by the backend from the event by accident.

RogueLogix replied:
yea, just making the record package-private is probably the better option, where backends have to implement their own.

RogueLogix replied:
records moved and made package-private

XFactHD commented on src/client/java/net/neoforged/neoforge/client/neo3d/validation/GpuDeviceUsageValidator.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    public void validatePipeline(RenderPipeline pipeline) {
+        if (pipeline.getColorLogic() != LogicOp.NONE && enabledFeatures.logicOp()) {
```
```suggestion
        if (pipeline.getColorLogic() != LogicOp.NONE && !enabledFeatures.logicOp()) {
```

XFactHD commented on src/client/java/net/neoforged/neoforge/client/neo3d/validation/GpuDeviceUsageValidator.java:
```diff
@@ -0,0 +1,134 @@
+        final int samplers = pipeline.getSamplers().size();
+        for (RenderPipeline.UniformDescription uniform : pipeline.getUniforms()) {
+            if (!properties.knownUniformTypes().contains(uniform.type())) {
+                throw new IllegalArgumentException(String.format("Unknown by device UniformType (%s) provided for uniform %s, known types are %s", uniform.type().name(), uniform.name(), Arrays.toString(properties.knownUniformTypes().toArray())));
```
```suggestion
                throw new IllegalArgumentException(String.format("Unknown by device UniformType (%s) provided for uniform %s, known types are %s", uniform.type().name(), uniform.name(), properties.knownUniformTypes()));
```
Converting the list or set to an array and then a string provides no benefit, same applies to all other cases where this is done.

RogueLogix replied:
It does here actually, this is specifically to ensure that the enum names are printed as an array, rather than the set object's hash.
the message currently prints something like `Unknown by device UniformType (SSBO) provided for uniform mySSBO, known types are [ UBO, UTB ]`

XFactHD replied:
`Set` and `EnumSet` implement `toString()` by virtue of `AbstractCollection` doing so and `Collections.UnmodifiableCollection` (which `Collections.UnmodifiableSet` extends) forwards the call to the inner set. There is no difference in output between using the set directly and converting it to an array and then a string.

RogueLogix replied:
i see, don't know how i missed that before. you should have write access to this branch if you want to fix it yourself, or i can get to it tonight.

XFactHD commented on src/client/java/net/neoforged/neoforge/client/neo3d/validation/GpuDeviceUsageValidator.java:
```diff
@@ -0,0 +1,134 @@
+        if (utbs + samplers > properties.maximumTextureBindings()) {
+            throw new IllegalArgumentException(String.format("Pipeline %s defines %s texture bindings, more than %s allowed", pipeline.getLocation(), utbs + samplers, properties.maximumTextureBindings()));
+        }
+        if (ubos + utbs + samplers > properties.maximumTextureBindings()) {
```
```suggestion
        if (ubos + utbs + samplers > properties.maximumUniformBindings()) {
```

XFactHD commented on src/client/java/net/neoforged/neoforge/client/neo3d/validation/GpuDeviceUsageValidator.java:
```diff
@@ -0,0 +1,134 @@
+                throw new IllegalArgumentException(String.format("Unknown by device SourceFactor (%s) provided for sourceAlpha, known factors are %s", blendFunc.sourceAlpha().name(), Arrays.toString(properties.knownSourceFactors().toArray())));
+            }
+            if (!properties.knownDestFactors().contains(blendFunc.destColor())) {
+                throw new IllegalArgumentException(String.format("Unknown by device SourceFactor (%s) provided for destColor, known factors are %s", blendFunc.destColor().name(), Arrays.toString(properties.knownDestFactors().toArray())));
```
```suggestion
                throw new IllegalArgumentException(String.format("Unknown by device DestFactor (%s) provided for destColor, known factors are %s", blendFunc.destColor().name(), Arrays.toString(properties.knownDestFactors().toArray())));
```

XFactHD commented on src/client/java/net/neoforged/neoforge/client/neo3d/validation/GpuDeviceUsageValidator.java:
```diff
@@ -0,0 +1,134 @@
+                throw new IllegalArgumentException(String.format("Unknown by device SourceFactor (%s) provided for destColor, known factors are %s", blendFunc.destColor().name(), Arrays.toString(properties.knownDestFactors().toArray())));
+            }
+            if (!properties.knownDestFactors().contains(blendFunc.destAlpha())) {
+                throw new IllegalArgumentException(String.format("Unknown by device SourceFactor (%s) provided for destAlpha, known factors are %s", blendFunc.destAlpha().name(), Arrays.toString(properties.knownDestFactors().toArray())));
```
```suggestion
                throw new IllegalArgumentException(String.format("Unknown by device DestFactor (%s) provided for destAlpha, known factors are %s", blendFunc.destAlpha().name(), Arrays.toString(properties.knownDestFactors().toArray())));
```

RogueLogix committed:
Apply suggestions from code review

Co-authored-by: Dennis C <xfacthd@gmx.de>

Technici4n commented on src/client/java/net/neoforged/neoforge/client/event/ConfigureGpuDeviceEvent.java:
```diff
@@ -0,0 +1,43 @@
+import net.neoforged.neoforge.client.neo3d.GpuDeviceFeatures;
+import net.neoforged.neoforge.client.neo3d.GpuDeviceProperties;
+
+public class ConfigureGpuDeviceEvent extends Event implements IModBusEvent, GpuDeviceFeatures {
```
This is still missing documentation.

RogueLogix replied:
added

Technici4n commented on src/client/java/net/neoforged/neoforge/client/event/ConfigureGpuDeviceEvent.java:
```diff
@@ -0,0 +1,43 @@
+import net.neoforged.neoforge.client.neo3d.GpuDeviceFeatures;
+import net.neoforged.neoforge.client.neo3d.GpuDeviceProperties;
+
+public class ConfigureGpuDeviceEvent extends Event implements IModBusEvent, GpuDeviceFeatures {
```
Implementing `GpuDeviceFeatures` is kinda useless here.

Technici4n replied:
I guess it's used to construct the immutable record. That is ok.

RogueLogix replied:
Allows it to be passed to the record constructor, this is done when making the record kept for the device lifetime (and returned by enabledFeatures) off of the ones enabled on the event.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceProperties.java:
```diff
@@ -0,0 +1,217 @@
+import java.util.Set;
+import org.jetbrains.annotations.ApiStatus;
+
+public interface GpuDeviceProperties {
```
I think we should only introduce properties that matter in the context of what we are extending. Most properties here can be removed as they don't achieve the goal of this PR (making Blaze3D extensible).

RogueLogix replied:
most of the ones added are covering existing B3D features, though they could be pulled from this PR and done in a second one instead if thats preferred.

Technici4n replied:
I would prefer that. Of course the properties that we are extending make sense to keep. Having that said we might want to merge the properties and the features into one interface, the difference is a bit fuzzy anyway.

RogueLogix replied:
The difference is that features are enableable, where properties is reporting stuff about it and there may be none or many for each feature. Notable that this pattern was pulled from how VK handles it. Properties are just limit reporting, while features is telling it what things you want to use.

Technici4n replied:
Non-standard `Set<DepthTestFunction> knownDepthTestFunctions();` will need to be enabled so they belong in features not properties. This is why I am saying that the distinction is quite fuzzy, and better not done at our level.

RogueLogix replied:
not-enabled and unknown are two different issues IMO. The "unknown" catch is there for a"this _will_ explode in prod" catch, while using something not-enabled may still work and should probably be a different validation error. _may still work_ is pointing more at the bitfields than the enums, where some are just hints anyway.

As another option, some things maybe we don't want to use explicit enablement and instead could rely on what is reported for if its enabled or not. Depth test functions seems like a perfect example of that where there is no need to enable them explicitly, just check if the backend knows about the one you want to use.

Technici4n replied:
They should be enabled explicitly to make sure that the modder is aware of the feature not potentially being available.

RogueLogix replied:
Fair enough, though maybe there should be second properties object that is created after device creation? Describing what is actually allowed to be used, rather than what the underlying backend is theoretically capable of. Properties as-is was mostly aimed at the latter case.

so, same interface, but the one you get from the created device will reflect that the extended depth tests weren't enabled, and therefore aren't allowed. The validation layer could also do strict enforcement on this,

Technici4n replied:
I think it makes sense to be able to query during the event whether the feature is available, and then later whether it was enabled. Same interface sounds fine. I'd maybe document clearly in the interface javadoc that it either represents the "potentially available" features or the enabled features depending on event usage vs. later usage?

RogueLogix replied:
besides interface javadoc, the two would also be gotten from very different places, that can also be named to imply that. ie: from the event is "maximalProperties" vs from the device "enabledProperties" _will need to find actually good names for that_

RogueLogix replied:
Most properties removed.

Comments reflect the intention of there being two different sets, however as there are no feature extensions included in this the same properties object (record) is still used in both circumstances.

Technici4n commented on patches/net/minecraft/client/renderer/CachedOrthoProjectionMatrixBuffer.java.patch:
```diff
@@ -0,0 +1,11 @@
+ 
+     private Matrix4f createProjectionMatrix(float p_418447_, float p_418426_) {
+-        return new Matrix4f().setOrtho(0.0F, p_418447_, this.invertY ? p_418426_ : 0.0F, this.invertY ? 0.0F : p_418426_, this.zNear, this.zFar);
++        return new Matrix4f().setOrtho(0.0F, p_418447_, this.invertY ? p_418426_ : 0.0F, this.invertY ? 0.0F : p_418426_, this.zNear, this.zFar, RenderSystem.getDevice().properties().depthZeroToOne());
```
This seems out of scope for this first PR tbh.

RogueLogix replied:
I'll pull and open a separate PR for it.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceFeatures.java:
```diff
@@ -0,0 +1,32 @@
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.neo3d;
```
The package name should be renamed to `blaze3d`, `neo` is already contained in earlier items of the package.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceFeatures.java:
```diff
@@ -0,0 +1,32 @@
+
+import org.jetbrains.annotations.ApiStatus;
+
+public interface GpuDeviceFeatures {
```
Missing javadoc

RogueLogix replied:
added

Technici4n commented on src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceProperties.java:
```diff
@@ -0,0 +1,217 @@
+import java.util.Set;
+import org.jetbrains.annotations.ApiStatus;
+
+public interface GpuDeviceProperties {
```
Missing javadoc

RogueLogix replied:
added

for docs in general, they could probably be done better than i did, im not the best at writing docs so suggestions welcome.

FiniteReality commented on src/client/java/net/neoforged/neoforge/client/event/ConfigureGpuDeviceEvent.java:
```diff
@@ -0,0 +1,43 @@
+            throw new UnsupportedOperationException("LogicOp is unavailable");
+        }
+        logicOp = true;
+    }
```
This should have a `requestExtension` and a `requireExtension` method which takes a string parameter.

Specifically, `requestExtension` would return whether the extension is available (and enabled, where not automatically enabled), and `requireExtension` would cause a mod loading error if the extension is not available.

marchermans replied:
Seems reasonable.

RogueLogix replied:
I believe this is outside the scope of what the API generic event should have, instead being implemented by the specific backend that an extension is being requested of.
Any usecase for this is likely to not just be API specific but even backend specific, as multiple for the same API are likely to have subtle behavior differences that may cause issues. To be able to account for that you should be aware not just of what API is being used but also exactly what backend. Directly asking a backend ensures that you are directly aware of the exact backend in use, not just the API its using and extensions prrovided.

RogueLogix replied:
Consensus was come to on discord that use of this is too backend specific, and should not be included as part of the generic interface.

RogueLogix committed:
rename neo3d package to blaze3d

RogueLogix committed:
move added GpuDevice functions to GpuDeviceExtension

RogueLogix committed:
make Immutable feature/properties records package-private and move to opengl package

RogueLogix committed:
Add missing license headers

RogueLogix committed:
Pull non-critical device properties, separate PR will be created later

RogueLogix committed:
Rename GpuDeviceExtension.properties to enabledProperties

RogueLogix committed:
Add javadoc

RogueLogix committed:
Use pattern matching if

RogueLogix committed:
Dont use `assert`

RogueLogix committed:
Add one-line comment for why an enum isn't exhaustive

RogueLogix committed:
Directly use AbstractCollection's toString

RogueLogix committed:
fix formatting

RogueLogix committed:
move enabledProperties and enabledFeatures impl to GlDevice

RogueLogix committed:
remove unused glCaps

RogueLogix commented:
Besides @FiniteReality's ask, I believe all other outstanding issues have been addressed. May still need some doc fixup, but code complete.

marchermans commented on src/client/java/net/neoforged/neoforge/client/ClientHooks.java:
```diff
@@ -1128,4 +1134,12 @@ public static List<PictureInPictureRendererRegistration<?>> gatherPictureInPictu
         return List.copyOf(vanillaRenderers);
     }
+
+    public static GpuDevice createGpuDevice(long window, int debugLevel, boolean syncDebug, BiFunction<ResourceLocation, ShaderType, String> defaultShaderSource, boolean enableDebugLabels) {
```
Missing JDoc

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceExtension.java:
```diff
@@ -0,0 +1,12 @@
+package net.neoforged.neoforge.client.blaze3d;
+
+public interface GpuDeviceExtension {
+    GpuDeviceProperties enabledProperties();
```
Missing JDoc

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceProperties.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    String apiVersionString();
+
+    int apiVersionMajor();
+
+    int apiVersionMinor();
```
Missing JDoc.

Does this always exist? For example how does Vulkan work here?

RogueLogix replied:
Vulkan would report major 1, minor 0-4, with the version string also having the 3rd patch number in it.

XFactHD commented on build.gradle:
```diff
@@ -75,6 +75,11 @@ immaculate {
+            if (fileContents.startsWith("--- a/com/mojang/blaze3d")) {
+                return fileContents
+            }
+            
```
The extension interface should be added via interface injection instead of special-casing B3D classes here.

XFactHD commented on patches/com/mojang/blaze3d/platform/DepthTestFunction.java.patch:
```diff
@@ -0,0 +1,10 @@
+ 
+ @OnlyIn(Dist.CLIENT)
+ @DontObfuscate
++@net.neoforged.neoforge.internal.NonExhaustiveEnum(reason = "Remaining depth test functions will be added")
```
I would prefer if this comment was worded a bit more generic and passive, i.e. along the lines of "Further depth test functions may be added". The same applies to the other affected enums.

RogueLogix committed:
Update NonExhaustiveEnum comments

RogueLogix committed:
Use interface injection for GpuDeviceExtension

XFactHD commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceProperties.java:
```diff
@@ -0,0 +1,105 @@
+    int knownGpuBufferUsageBits();
+
+    /**
+     * @see GpuDeviceProperties#knownGpuTextureUsageBits()
```
```suggestion
     * @see GpuDeviceProperties#knownGpuBufferUsageBits()
```

XFactHD commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceProperties.java:
```diff
@@ -0,0 +1,105 @@
+    /**
+     * All sets returned must be unmodifiable, and should not be unique instances
+     * <br>
+     * Backend are recommended to use {@link EnumSet#of()} rather than {@link EnumSet#range(Enum, Enum)} to ensure that only known values are returned
+     * regardless of where future enum additions are added.
+     * Use of {@link EnumSet#range(Enum, Enum)} may in an enum added between those that did not previously exist being considered known.
```
```suggestion
     * Backends are recommended to use {@link EnumSet#of()} rather than {@link EnumSet#range(Enum, Enum)} to ensure that only known values are returned
     * regardless of where future enum additions are added.
     * Use of {@link EnumSet#range(Enum, Enum)} may include an enum constant added between those that did not previously exist being considered known.
```

XFactHD commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceProperties.java:
```diff
@@ -0,0 +1,105 @@
+     * <br>
+     * Bits or enums unknown by the backend must not be used. These may be defined by a newer NeoForge version than the backend was built against, or the feature may not be enabled.
+     * <br>
+     * Backends must not return known values for features that are not enabled, even if they are supported.
```
```suggestion
     * Backends must not return known values as enabled for features that are not enabled, even if they are supported.
```

XFactHD reviewed (changes requested):
Two general points:
- Newly added files should use the NeoForge license header, not the old MCForge one
- Make sure sentences in javadocs end with a period

RogueLogix committed:
Apply suggestions from code review

Co-authored-by: Dennis C <xfacthd@gmx.de>

RogueLogix committed:
Use correct license header for new files

RogueLogix committed:
Docs fixup

RogueLogix committed:
Immaculate

RogueLogix committed:
Add property to disable ValidationGpuDevice

marchermans commented on src/client/java/net/neoforged/neoforge/client/event/ConfigureGpuDeviceEvent.java:
```diff
@@ -0,0 +1,59 @@
+        return logicOp;
+    }
+
+    public void enableLogicOp() {
```
Missing JDoc

marchermans commented on src/client/java/net/neoforged/neoforge/client/loading/NeoForgeLoadingOverlay.java:
```diff
@@ -52,7 +53,11 @@ public NeoForgeLoadingOverlay(final Minecraft mc, final ReloadInstance reloader,
-        var framebuffer = ((GlDevice) RenderSystem.getDevice()).createExternalTexture("loading overlay framebuffer", GpuTexture.USAGE_TEXTURE_BINDING, displayWindow.getFramebufferTextureId());
+        var gpuDevice = RenderSystem.getDevice();
+        if (gpuDevice instanceof ValidationGpuDevice validationGpuDevice) {
+            gpuDevice = validationGpuDevice.getRealDevice();
```
Leave a comment why you need the real device here.

marchermans commented on src/client/java/net/neoforged/neoforge/client/loading/NeoForgeLoadingOverlay.java:
```diff
@@ -110,7 +115,11 @@ static class ExternalTexture extends AbstractTexture {
-            this.textureView = RenderSystem.getDevice().createTextureView(texture);
+            var gpuDevice = RenderSystem.getDevice();
+            if (gpuDevice instanceof ValidationGpuDevice validationGpuDevice) {
+                gpuDevice = validationGpuDevice.getRealDevice();
```
Here as well. Even if it is just temporary.

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceExtension.java:
```diff
@@ -0,0 +1,12 @@
+public interface GpuDeviceExtension {
+    GpuDeviceProperties enabledProperties();
+
+    GpuDeviceFeatures enabledFeatures();
```
Missing JDoc

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/validation/ValidationRenderPass.java:
```diff
@@ -0,0 +1,108 @@
+import net.neoforged.neoforge.client.stencil.StencilTest;
+import org.jetbrains.annotations.Nullable;
+
+public class ValidationRenderPass implements RenderPass {
```
Provide documentation on what this does any why. Counts for the entire class.

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/validation/ValidationGpuTexture.java:
```diff
@@ -0,0 +1,58 @@
+import com.mojang.blaze3d.textures.GpuTexture;
+import org.jetbrains.annotations.ApiStatus;
+
+public class ValidationGpuTexture extends GpuTexture {
```
Docs

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/validation/ValidationGpuDevice.java:
```diff
@@ -0,0 +1,173 @@
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+public class ValidationGpuDevice implements GpuDevice {
```
Docs

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/validation/ValidationCommandEncoder.java:
```diff
@@ -0,0 +1,172 @@
+import java.util.function.Supplier;
+import org.jetbrains.annotations.Nullable;
+
+public class ValidationCommandEncoder implements CommandEncoder {
```
Docs

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/validation/GpuDeviceUsageValidator.java:
```diff
@@ -0,0 +1,78 @@
+import net.neoforged.neoforge.client.blaze3d.GpuDeviceFeatures;
+import net.neoforged.neoforge.client.blaze3d.GpuDeviceProperties;
+
+public class GpuDeviceUsageValidator {
```
Docs

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/opengl/NeoGlDeviceProperties.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.Set;
+import net.neoforged.neoforge.client.blaze3d.GpuDeviceProperties;
+
+public class NeoGlDeviceProperties implements GpuDeviceProperties {
```
Docs

marchermans commented on src/client/java/net/neoforged/neoforge/client/blaze3d/opengl/NeoGlDeviceProperties.java:
```diff
@@ -0,0 +1,96 @@
+public class NeoGlDeviceProperties implements GpuDeviceProperties {
+    @Override
+    public String backendName() {
+        return "Default";
```
Either name the class `DefaultOpenGLDeviceProperties` or rename this backend

RogueLogix committed:
Address docs/naming feedback.

RogueLogix committed:
formatting

RogueLogix committed:
Remove underlying API version properties. This information shouldn't be necessary unless you are going around the B3D backend.

RogueLogix committed:
Fix GlDevice patch

XFactHD commented on src/client/java/net/neoforged/neoforge/client/ClientHooks.java:
```diff
@@ -1128,4 +1134,17 @@ public static List<PictureInPictureRendererRegistration<?>> gatherPictureInPictu
+    /**
+     * Creates Blaze3d backend GpuDevice, and potentially wraps it in a validation layer.
+     * <br>
+     * Alternate backends should target this function with mixins instead of {@link RenderSystem#initRenderer(long, int, boolean, BiFunction, boolean)}.
```
How is an alternate backend supposed to target this with a mixin? The `glDevice` variable will be of type `GlDevice` due to the use of `var`, so they won't be able to assign a different `GpuDevice` implementation to the variable.

RogueLogix replied:
Cancelable inject at head.

XFactHD replied:
What's the benefit over targetting the device instantiation in `RenderSystem.initRenderer()` then?

RogueLogix replied:
Fallback to other backends or Neo's default without special handling.

XFactHD replied:
That can be done equally trivial at the `RenderSystem` level using `@WrapOperation`

RogueLogix replied:
This was also done at first to pull the `ValidationGpuDevice` creation out of a patch, the mixin targeting being nicer (without mixin extras) was a secondary benefit.

XFactHD replied:
There is absolutely no reason to avoid MixinExtras for this, it's shipped by Neo and always available. Having to effectively overwrite the method (a cancellable inject that's always cancelled is realistically nothing more than a silent overwrite) is not a nicer target in my opinion (not that a `@WrapOperation` that never calls its original op is any better, to be fair).

If there should be any reason for a custom backend to target the `ClientHooks` method, then it should be the benefit of just having to replace the device instantiation and leaving everything else, including the validation device wrapping, to Neo. However, targetting the device instantation in the `ClientHooks` method would require the actual instantation to be moved to yet another method, the call to which would then be targetted, as the `@WrapOperation`'s return type must exactly match the type being instantiated or returned (super-type doesn't work). Actually doing this seems pretty ridiculous to me though.

All this, just to say: the recommendation made by the doc comment is pretty pointless and may as well not exist.

RogueLogix replied:
I'm just going to yeet the whole javadoc then, it was only added because [Orion asked for it](https://github.com/neoforged/NeoForge/pull/2399#discussion_r2197762176), but I'm also not leaving a doc that says the same thing the 4 lines of code following it do.

The validation wrapping can't just be left to Neo in all cases though. Cinnabar will be adding its own extensions that it needs to validate the usage of, and therefore will also need to extend the ValidationGpuDevice and handle doing the wrapping itself.

That second line is mostly saying _don't do what I [Cinnabar] am currently doing with an Overwrite on `RenderSystem#initRenderer`_ so its mostly moot in the first place.

RogueLogix committed:
Address PR feedback

Technici4n commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceExtension.java:
```diff
@@ -0,0 +1,26 @@
+
+import net.neoforged.neoforge.client.event.ConfigureGpuDeviceEvent;
+
+public interface GpuDeviceExtension {
```
We should probably move this to `net.neoforged.neoforge.client.extension` for consistency with our other extensions.

Technici4n commented on src/main/java/net/neoforged/neoforge/internal/NonExhaustiveEnum.java:
```diff
@@ -0,0 +1,30 @@
+
+/**
+ * Signifies that an enum is not exhaustive, and additional values may be added while NeoForge is stable.
+ * <br>
```
Nit: Avoid using `<br>` for javadoc. New paragraphs are preferred.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceFeatures.java:
```diff
@@ -0,0 +1,21 @@
+
+/**
+ * Enable-able features of a Blaze3d backend GpuDevice.
+ * <br>
```
Nit: Use `<p>` not `<br>`.

Technici4n replied:
Nit: Use `<p>` not `<br>`.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceFeatures.java:
```diff
@@ -0,0 +1,21 @@
+package net.neoforged.neoforge.client.blaze3d;
+
+/**
+ * Enable-able features of a Blaze3d backend GpuDevice.
```
There is really no difference compared to properties, which should also be enabled during configuration. So let's merge the two interfaces.

Technici4n replied:
This was further discussed on Discord since, see relevant discussion there. Leaving the comment to remember to make the changes.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/blaze3d/GpuDeviceProperties.java:
```diff
@@ -0,0 +1,94 @@
+     * <br>
+     * Backends must not return known values as enabled for features that are not enabled, even if they are supported.
+     */
+    int knownGpuBufferUsageBits();
+
+    /**
+     * @see GpuDeviceProperties#knownGpuBufferUsageBits()
+     */
+    int knownGpuTextureUsageBits();
```
Do we need to mark the bits as "non-exhaustive" as well then?

RogueLogix replied:
i would say no because unlike with the where new values are getting added, the extra bits already exist in the `int` type.

these also dont have anywhere that would be good to actually define the non-exhaustive on, because they are just free `public static final int`s in the GpuBuffer and GpuTexture classes

Technici4n commented on src/client/java/net/neoforged/neoforge/client/blaze3d/validation/ValidationGpuDevice.java:
```diff
@@ -0,0 +1,184 @@
+/**
+ * The validation GpuDevice is used to ensure that both mods and backends are complying with the B3D API contract correctly.
+ * <p>
+ * Validation done by this class includes:
```
Nit: Use a proper html list.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/blaze3d/validation/GpuDeviceUsageValidator.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    public void validatePipeline(RenderPipeline pipeline) {
```
Is there a reason for this one method to be public? If yes, it would be worth adding a comment maybe?

RogueLogix replied:
no, it should be package private

Technici4n commented on src/client/java/net/neoforged/neoforge/client/blaze3d/validation/ValidationGpuTexture.java:
```diff
@@ -0,0 +1,65 @@
+
+/**
+ * Validation wrapper around GpuTexture
+ * <br>
```
Nit: Don't use `<br>`.

Technici4n commented on src/client/java/net/neoforged/neoforge/client/ClientHooks.java:
```diff
@@ -1128,4 +1134,12 @@ public static List<PictureInPictureRendererRegistration<?>> gatherPictureInPictu
+
+    public static GpuDevice createGpuDevice(long window, int debugLevel, boolean syncDebug, BiFunction<ResourceLocation, ShaderType, String> defaultShaderSource, boolean enableDebugLabels) {
+        final var glDevice = new GlDevice(window, debugLevel, syncDebug, defaultShaderSource, enableDebugLabels);
+        if (FMLLoader.isProduction() || Boolean.getBoolean("neoforge.disableB3DValidation")) {
```
Idea: make this a config option instead, with a default to `!FMLLoader.isProduction()`, such that it could be overridden both in dev and in prod.

RogueLogix committed:
Remove known* properties in favor of validation layer itself knowing
device properties are also now constant, and may report properties for disabled features.

RogueLogix committed:
Use config value to enable B3D validation layer.

RogueLogix committed:
Move GpuDeviceExtension to extensions package

Technici4n commented on src/main/resources/META-INF/injected-interfaces.json:
```diff
@@ -302,5 +302,8 @@
   "net/minecraft/world/level/biome/BiomeSpecialEffects$GrassColorModifier": [
     "net/neoforged/fml/common/asm/enumextension/IExtensibleEnum"
+  ],
+  "com/mojang/blaze3d/systems/GpuDevice": [
```
Let's keep this addition sorted - even though some weren't. (Since I am requesting another change anyway).

Technici4n commented on src/main/resources/assets/neoforge/lang/en_us.json:
```diff
@@ -194,6 +194,7 @@
   "neoforge.configgui.advertiseDedicatedServerToLan.tooltip": "Set this to true to enable advertising the dedicated server to local LAN clients so that it shows up in the Multiplayer screen automatically.",
   "neoforge.configgui.attributeAdvancedTooltipDebugInfo": "Additional Attribute Advanced Tooltips",
   "neoforge.configgui.attributeAdvancedTooltipDebugInfo.tooltip": "Set this to true to enable additional information about attributes on an item when advanced tooltips is on.",
+  "neoforge.configgui.enableB3DValidationLayer": "When enabled, all usage of B3D will be validated against allowed usage",
```
This should also have a tooltip version for the config GUI. I would also avoid using the B3D abbreviation as this might be unfamiliar to random users.

For example:

```suggestion
  "neoforge.configgui.enableB3DValidation": "Enable Blaze3D Validation Layer",
  "neoforge.configgui.enableB3DValidation.tooltip": "When enabled, all usage of Blaze3D (Minecraft's internal rendering abstraction) will be validated against allowed usage.",
```

Technici4n reviewed (changes requested):
LGTM. Just two minor requests.

RogueLogix committed:
Update src/main/resources/assets/neoforge/lang/en_us.json

Co-authored-by: Bruno Ploumhans <13494793+Technici4n@users.noreply.github.com>

RogueLogix committed:
Sort interface injection

RogueLogix committed:
Use HTML list for jdoc

RogueLogix committed:
fix en_us enableB3DValidationLayer key
